### ISSUE TITLE
Multiple linked 'Clients' (like emacs 'frames') [enhancement] [idea]

### ISSUE THREAD

**Issue by ChrisPenner** [neovim/neovim#2161]
Hi Everyone, I've got a few use-cases that I'm wondering if Neovim can solve. If I'm missing something just let me know.

The main thing I'd like to do is to be able to have two or more *linked* Windows (using OS terminology, NOT vim windows, see emacs *frames* for a comparison) of Vim. The reason for this is that it would facilitate multi-monitor set-ups. Each monitor would have its own OS window (I'll call them *widgets* from now on to avoid confusion) and focus could be switched between the two (or more) using vim-bindings.

This is a necessary addition to existing structures like tabs, windows, splits, etc. because often different monitors have different resolutions and arrangements, and Vim doesn't always adapt well to these situations (stretching a single widget across multiple screens doesn't work well, and MANY programmers are using multi-monitor setups).

Some features this would require for optimal use would include:

* If a buffer is present in each widget, they should update themselves with changes made in other widgets (preferably in real-time) 
* It is possible for a buffer to be open in one widget, while not open in another.
* It is possible for a buffer to be open in multiple widgets at the same time.
* All splits, windows, and tabs are managed on a per widget basis (making a split in one widget doesn't affect any others).
* There is only one active cursor between all widgets
* The Vim instance is aware of all windows and can switch between them using keyboard bindings (similar to switching between splits using <ctrl-w>[key] bindings

As far as implementation goes I don't believe something like this is currently possible. Since Vim does not traditionally deal with Widgets itself (leaving that to gVim and MacVim) it would perhaps be more dependent on GUI clients to implement, however it would still be important for Neovim to define an interface which allows this behaviour. These features will come in handy when third-parties begin to look at embedding Vim, and will allow them greater flexibility in doing so.

I don't believe that Vim's server modes allow this behaviour as they are, though it would be possible to extend it's behaviour to accommodate this sort of thing.

If there is any confusion, please ask.  For reference, simply look over how "frames" are handled in Emacs.

Cheers! Thanks for your time.

**Comment by tarruda** [comment]
Something I had planned is to allow one nvim instance have multiple screens(possibly of different sizes) attached. The main purpose is to allow more fancy GUIs(which have panes with different font sizes for example), but also covers the use case you described.

**Comment by ChrisPenner** [comment]
Right, that sounds good. I think GUI providers will find many convenient uses for a design like this.

Would it be up to the GUI provider to give suitable key-bindings to navigate between them? Or is it possible to have a vim-centric solution for focusing a screen using vim keybindings?
Using OS shortcuts like Alt-Tab is kind of a sub-optimal solution.

**Comment by justinmk** [comment]
@ChrisPenner those are details which will need to be discussed when the time comes to focus on this.

**Comment by simonweil** [comment]
:+1:

**Comment by teto** [comment]
I am also very interested in this.

Meanwhile if your main interest is using the same set of command to move between splits as your window manager, you should be able to use the rpc to acheive that.

I was thinking of writing a windowmanager (i3 in my case) level script that would detect neovim focus and then send wincmd via rpc. When the command fails (i.e. you run wincmd h and there is no window at the left), the wm can takeover and switch focus to the window on the left.

My current problem is to retrieve the "wincmd h" result.

**Comment by teto** [comment]
I've written and explained a hack to move smoothly (i.e., with the same mappings) between my tiling Window Manager windows and neovim splits in this post: https://www.reddit.com/r/vim/comments/3hh42z/can_i_replace_vims_windows_with_a_tiling_window/cuc8gk5
That doesn't solve everything but that helps

**Comment by languitar** [comment]
:+1:

**Comment by teto** [comment]
I think it should be possible through a plugin to hack something with Shada, like:
-on some events  you force shada to save (yanking/focuslost...) or load shada file (gainedfocus)
-and you override some commands like :split / newbuffer/new tab to open a new neovim instance.

**Comment by Tarmean** [comment]
I guess flushing everything to disk or using the same swapfiles could make this work but I am not sure if that would be a good idea.

Does focuslost/gainedfocus even work in the terminal version? Haven't tried much with gui clients because so far all I tried very immensely laggy compared to terminal nvim.

**Comment by teto** [comment]
Using the disk is indeed not optimal, hence the "hack" part :) 
focuslost/gainedfocus should be working in the terminal version since some patches last week (haven't tested yet).
That would be a cool hack though

**Comment by teto** [comment]
I converted my post from the 23 August into an actual script that can be used in i3 to use the same WM bindings to move between splits and windows, https://github.com/teto/i3-dispatch .
This doesn't solve the other problems but I find it useful already.

**Comment by haasn** [comment]
Big +1 on this. It's something I've always wished for.

For even bigger bonus the split-style commands (e.g. :Sex or :stag) should just pop up a new terminal emulator with a vim instance sitting inside it. My window manager will take care of the rest. :)

**Comment by jreybert** [comment]
I hope one day to be able to use vim in the same project on two monitors!

BTW, if you know any workaround for the moment, it would be highly appreciated.

**Comment by vatrat** [comment]
Just as a note for usability, it might make sense to have each OS window show a vim tab, like in XMonad. Having both windows open the same tab could just mirror the windows, maybe with the cursor only existing in one window, OR trying to open a tab that is currently open in another window could just swap the remote window and the current window's tabs. Like this:
[(window1, tab1), (window2, tab2)]
*in window1, run a command to open next tab, or open tab2 specifically, or whatever*
[(window1,tab2), (window2, tab1)]
Tab2 is opened as requested, and the previously open tab is moved to fill the gap of window2. This is how XMonad deals with this problem, and I think it could work here, as well.

**Comment by stephane-archer** [comment]
Dear @jreybert,
I have the same dream as you and I find a solution.... EmacsEvilMode.
It's hard to say but Emacs is a better Vim than Vim itself ;(
Use an editor on multiple monitor should be a basic feature.
I really hope that NeoVim will come with this extremely useful feature :)

**Comment by haasn** [comment]
> Just as a note for usability, it might make sense to have each OS window show a vim tab, like in XMonad. Having both windows open the same tab could just mirror the windows, maybe with the cursor only existing in one window, OR trying to open a tab that is currently open in another window could just swap the remote window and the current window's tabs. Like this:

Of course, if every tab was literally its own window, xmonad etc. would deal with this just fine. :p

**Comment by stephane-archer** [comment]
Sorry @vatrat , I didn´t understand how you solve the problem :( I use i3wm witch look like close to XMonad. In my case I want to edit a C++ file on one screen and have on another screen some file to help me (like the header or the declaration of over function inside the same file). Can it can solve my problem?

**Comment by bfredl** [comment]
I think this can be done (single nvim instance solution) without gigantic refactors or additions to screen.c and window.c with one sacrifice: top-level frames will replace tabs in multi-frame mode. I have a WIP branch here https://github.com/neovim/neovim/compare/master...bfredl:multigrid?expand=1 . All tab functions and commands will manipulate top-level frames instead in multigrid mode. Currently can be "tested" only with https://github.com/neovim/python-gui/compare/master...bfredl:multigrid?expand=1 . I will open PR when I fix some severe problems such as no input focus tracking and almost completely broken resize. Also it must be compatible with floating windows #6619 using own grids.

**Comment by lorienhu** [comment]
[There is a bounty on this issue.](https://www.bountysource.com/issues/9598868-multiple-linked-clients-like-frames-in-emacs)

**Comment by kovasap** [comment]
any updates on this issue?

**Comment by bfredl** [comment]
@kovasap #8455 and #6619 are adding infrastructure that lays the groundwork for tackling this. After those are merged we should move forward with this. #8707 will (among a lot of other things) allow UI-side implementation of frames, while #7541 is a rough sketch of a solution with nvim-side control (using the magic formula "one tabpage == one frame").

**Comment by kovasap** [comment]
Ok, thanks for the quick and detailed update!

**Comment by the42** [comment]
long time no hear, eager to see this coming. May this also be the foundation for collaborative editing?

**Comment by teto** [comment]
There is work in that direction, for instance https://github.com/neovim/neovim/pull/11326

**Comment by dseeni** [comment]
gimme gimme would be nice, esp since we have a terminal too

**Comment by narodnik** [comment]
Amazing work guys!

**Comment by narodnik** [comment]
Waiting ages for this in vim. Love NeoVim

**Comment by mark-i-m** [comment]
Yes, thank you! Any updates since last year?

**Comment by MRAAGH** [comment]
Personally I have little interest in using vim bindings for controlling my window manager.

The part that's actually annoying is that when I have Vim open on two screens, I need to be careful not to open the same file in both of them.

Another inconvenience is that I don't get autocompletion from the other instance. Only the buffers in the current window are used.
